core/src/Kokkos_MathematicalFunctions.hpp

```diff
@@ -55,124 +55,233 @@
+  KOKKOS_INLINE_FUNCTION float FUNC##f(float x, float y) {              \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
+    return FUNC(x, y);                                                  \
+  }                                                                     \
+  template <class T1, class T2>                                         \
+  KOKKOS_INLINE_FUNCTION std::enable_if_t<                              \
+      std::is_arithmetic<T1>::value && std::is_arithmetic<T2>::value && \
+          !std::is_same<T1, long double>::value &&                      \
+          !std::is_same<T2, long double>::value,                        \
+      Kokkos::Impl::promote_2_t<T1, T2>>                                \
+  FUNC(T1 x, T2 y) {                                                    \
+    using Promoted = Kokkos::Impl::promote_2_t<T1, T2>;                 \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
+    return FUNC(static_cast<Promoted>(x), static_cast<Promoted>(y));    \
+  }
 
 #endif
 
 // Basic operations
+// clang-format off
+#ifdef KOKKOS_ENABLE_SYCL
+KOKKOS_INLINE_FUNCTION int abs(int n) { return sycl::abs(n); }
+KOKKOS_INLINE_FUNCTION long abs(long n) { return sycl::abs(n); }
+KOKKOS_INLINE_FUNCTION long long abs(long long n) { return sycl::abs(n); }
+// NOTE sycl::abs(genfloat) -> genfloat math function not available
+KOKKOS_INLINE_FUNCTION float abs(float x) { return sycl::fabs(x); }
+KOKKOS_INLINE_FUNCTION double abs(double x) { return sycl::fabs(x); }
+#else
+KOKKOS_INLINE_FUNCTION int abs(int n) { using std::abs; return abs(n); }
+KOKKOS_INLINE_FUNCTION long abs(long n) { using std::abs; return abs(n); }
+KOKKOS_INLINE_FUNCTION long long abs(long long n) { using std::abs; return abs(n); }
+KOKKOS_INLINE_FUNCTION float abs(float x) { using std::abs; return abs(x); }
+KOKKOS_INLINE_FUNCTION double abs(double x) { using std::abs; return abs(x); }
+#if !(defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) || defined(KOKKOS_ENABLE_OPENMPTARGET))
+KOKKOS_INLINE_FUNCTION long double abs(long double x) { using std::abs; return abs(x); }
+#endif
+#endif
+// clang-format on
 KOKKOS_IMPL_MATH_UNARY_FUNCTION(fabs)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmod)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(remainder)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmin)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmax)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(fdim)
 #ifndef KOKKOS_ENABLE_SYCL
-KOKKOS_IMPL_MATH_NAN()
+KOKKOS_INLINE_FUNCTION float nanf(char const* arg) { return ::nanf(arg); }
+KOKKOS_INLINE_FUNCTION double nan(char const* arg) { return ::nan(arg); }
+#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
+KOKKOS_INLINE_FUNCTION long double nanl(char const* arg) { return ::nanl(arg); }
+#endif
+#else
+// FIXME_SYCL
+KOKKOS_INLINE_FUNCTION float nanf(char const*) { return sycl::nan(0u); }
+KOKKOS_INLINE_FUNCTION double nan(char const*) { return sycl::nan(0ul); }
```

Comment:
What's the issue here? What's the problem with passing `arg`?
Would you please add a comment?

Reply:
`sycl::nan` does not follow the C/C++ standard library and takes an unsigned integer as argument.